src/probability_theory/stopping.lean

```diff
@@ -306,4 +328,121 @@ end linear_order
+λ x, u (τ x) x
+
+variable [linear_order ι]
+
+/-- Given a map `u : ι → α → E`, the stopped process with respect to `τ` is `u i x` if
+`i ≤ τ x`, and `u (τ x) x` otherwise.
+
+Intuitively, the stopped process stops evolving once the stopping time has occured. -/
+def stopped_process (u : ι → α → β) (τ : α → ι) : ι → α → β :=
+λ i x, u (linear_order.min i (τ x)) x
+
+lemma stopped_process_eq_of_le {u : ι → α → β} {τ : α → ι}
+  {i : ι} {x : α} (h : i ≤ τ x) : stopped_process u τ i x = u i x :=
+by simp [stopped_process, min_eq_left h]
+
+lemma stopped_process_eq_of_ge {u : ι → α → β} {τ : α → ι}
+  {i : ι} {x : α} (h : τ x ≤ i) : stopped_process u τ i x = u (τ x) x :=
+by simp [stopped_process, min_eq_right h]
+
+-- We will need cadlag to generalize the following to continuous processes
+section nat
+
+open filtration
+
+variables {f : filtration ℕ m} {u : ℕ → α → β} {τ : α → ℕ}
+
+section add_comm_monoid
+
+variables [add_comm_monoid β]
+
+lemma stopped_process_eq (n : ℕ) :
+  stopped_process u τ n =
+  set.indicator {a | n ≤ τ a} (u n) +
+    ∑ i in finset.range n, set.indicator {a | τ a = i} (u i) :=
+begin
+  ext x,
+  rw [pi.add_apply, finset.sum_apply],
+  cases le_or_lt n (τ x),
+  { rw [stopped_process_eq_of_le h, set.indicator_of_mem, finset.sum_eq_zero, add_zero],
+    { intros m hm,
+      rw finset.mem_range at hm,
+      exact set.indicator_of_not_mem ((lt_of_lt_of_le hm h).ne.symm) _ },
+    { exact h } },
+  { rw [stopped_process_eq_of_ge (le_of_lt h), finset.sum_eq_single_of_mem (τ x)],
+    { rw [set.indicator_of_not_mem, zero_add, set.indicator_of_mem],
+      { exact rfl }, -- refl does not work
+      { exact not_le.2 h } },
+    { rwa [finset.mem_range] },
+    { intros b hb hneq,
+      rw set.indicator_of_not_mem,
+      exact hneq.symm } },
+end
+
+lemma adapted.stopped_process_adapted [measurable_space β] [has_measurable_add₂ β]
+  (hu : adapted f u) (hτ : is_stopping_time f τ) :
+  adapted f (stopped_process u τ) :=
+begin
+  intro i,
+  rw stopped_process_eq,
+  refine @measurable.add _ _ _ _ (f i) _ _ _ _ _,
+  { refine @measurable.indicator _ _ (f i) _ _ _ _ (hu i) _,
+    convert @measurable_set.union _ (f i) _ _
+      (@measurable_set.compl _ _ (f i) (hτ i)) (hτ.measurable_set_eq i),
+    ext x,
+    change i ≤ τ x ↔ ¬ τ x ≤ i ∨ τ x = i,
+    rw [not_le, le_iff_lt_or_eq, eq_comm] },
+  { refine @finset.measurable_sum' _ _ _ _ _ _ (f i) _ _ _,
+    refine λ j hij, @measurable.indicator _ _ (f i) _ _ _ _ _ _,
+    { rw finset.mem_range at hij,
+      exact measurable.le (f.mono hij.le) (hu j) },
+    { rw finset.mem_range at hij,
+      refine f.mono hij.le _ _,
+      convert hτ.measurable_set_eq j, } }
+end
+
+end add_comm_monoid
+
+section normed_group
+
+variables [measurable_space β] [normed_group β] [has_measurable_add₂ β]
+
+lemma measurable_stopped_process (hτ : is_stopping_time f τ) (hu₁ : adapted f u) (n : ℕ) :
+  measurable (stopped_process u τ n) :=
+(hu₁.stopped_process_adapted hτ n).le (f.le _)
```

Comment:
the linter may complain about an unused argument here. I'll fix it if that's the case.